File: Scripts/generated_codes/java/17.java
"""
"""

# Enter your code here. Read input from STDIN. Print output to STDOUT
from collections import defaultdict
from itertools import combinations

n_cases = int(input())
for _ in range(n_cases):
    n, k, e, m = map(int, input().split())
    scores = []
    for _ in range(n-1):
        scores.append(list(map(int, input().split())))
    sergey_scores = list(map(int, input().split()))
    total_scores = []
    for i in range(n-1):
        total_scores.append(sum(scores[i][:-1]) + sergey_scores[i])
    total_scores.sort()
    n_students_in_university = n-1-k
    required_score = total_scores[n_students_in_university]
    required_score = -1 * required_score
    required_score += 1
    required_score += m
    print(required_score)

File: Scripts/generated_codes/python/codechef_responses/17.py
"""
"""

t = int(input())
while t > 0:
    n, k, e, m = map(int, input().split())
    a = []
    for i in range(n - 1):
        a.append(list(map(int, input().split())))
    a = sorted(a, reverse=True, key=lambda x: sum(x[:e - 1]))
    if sum(a[k - 1][:e - 1]) + 1 < sum(map(int, input().split())) + m:
        print(sum(map(int, input().split())) + m - sum(a[k - 1][:e - 1]))
    else:
        print("Impossible")
    t -= 1